akka-http-core/src/main/scala/akka/http/impl/engine/rendering/HttpResponseRendererFactory.scala

```diff
@@ -233,8 +233,21 @@ private[http] class HttpResponseRendererFactory(
               r ~~ `Transfer-Encoding` ~~ ChunkedBytes ~~ CrLf
           }
 
-          def renderContentLengthHeader(contentLength: Long) =
-            if (status.allowsEntity) r ~~ ContentLengthBytes ~~ contentLength ~~ CrLf else r
+          def renderContentLengthHeader(contentLength: Long) = {
+            // evaluate requirements for content-length according to https://httpwg.org/specs/rfc9110.html#field.content-length
+            // - for HEAD it is technically allowed, but must match the content-length of hypothetical GET request, which is not handled here
+            // - for 304 (Not Modified) it must similarly match the content-length of hypothetical 200-accepted request, which is not handled here
+            // - for 1xx (Informational) or 204 (No Content) it is explicitly not allowed
+            // - for 2xx (Successful) it is explicitly not allowed when the method is CONNECT
+            val isContentLengthForbidden =
+              (ctx.requestMethod eq HttpMethods.HEAD) ||
+              (status eq StatusCodes.NotModified) ||
+              (100 to 199 contains status.intValue) ||
+              (status eq StatusCodes.NoContent) ||
+              ((ctx.requestMethod eq HttpMethods.CONNECT) && (200 to 299 contains status.intValue))
```

Comment:
Since this is called on every request-response we need to be careful to make it as cheap as possible for the most common cases (and if for all cases even better). Comparing with a range allocates a `scala.util.Range` each time for example.

I wonder if we could push it to a call on `StatusCode` and maybe `HttpMethod` as well (something like `Method#contentLengthAllowed(forStatus: StatusCode)`) instead, making it cheaper.

Reply:
Sounds good, I added it to HttpMethod :+1: